### ISSUE TITLE
Implement ChEES-HMC

### ISSUE THREAD

**Issue by albcab** [blackjax-devs/blackjax#382]
I have a working draft of ChEES-HMC, except its not JAX-ready yet. The "Select jittered trajectory length..." line of [Algorithm 1](http://proceedings.mlr.press/v130/hoffman21a/hoffman21a.pdf) requires a float from a predetermined Halton sequence of length wamup+sampling number of iterations. When returning the kernel after warm-up, it needs to keep yielding from the sequence at every iteration (but JAX functions need to be pure).

For now my implementation yields the next number using a scipy:
```python
halton_sequence_gn = scipy.stats.qmc.Halton(d=1)
#then inside the kernel...
currrent_halton_number = jnp.array(halton_sequence_gn.random(n=1).squeeze())
```
Any help is appreciated. Will keep updated on progress.

**Comment by rlouf** [comment]
Afair the Halton sequence is a deterministic function of a base $b$ and an index $i$, correct? In this case we can initialize the kernel with a function that returns the number of the sequence at index $i$. From here I see two choices:

1. The sequence itself matters (we need to return subsequent numbers) in which case we can keep track of the value of $i$ in `ChEEsState` and increment it at every step.
2. We use the integers contained in an `rng_key` to generate the number from the Halton sequence.

**Comment by albcab** [comment]
I was considering creating an extended state like
```python
class ChESSState(NamedTuple):
    state: hmc.HMCState
    current_iteration: int
```
but the second solution is much, much nicer. I don't think the order matters, scrambling the sequence seems to be a positive thing even. I'll study Halton sequences more and look into generating it from the `rng_key` :+1: :+1: :+1:

**Comment by rlouf** [comment]
> How about we use the `tfp.mcmc.sample_halton_sequence` from tensorflow_probability https://www.tensorflow.org/probability/api_docs/python/tfp/substrates/jax/mcmc/sample_halton_sequence?

Let's avoid dependencies on another library (especially heavyweight libraries) whenever we can!

**Comment by albcab** [comment]
A Halton sequence shouldn't be too complicated to implement.

**Comment by rlouf** [comment]
> but the second solution is much, much nicer. 

There is something very satisfying in using PRNGKeys to generate this pseudo-random sequence indeed.

**Comment by albcab** [comment]
> A Halton sequence shouldn't be too complicated to implement.

Famous last words.

The order does seem to matter for the Halton sequence. Even if scrambling can be necessary, how the scrambling is done does matter (see [Owen, 2017](https://arxiv.org/pdf/1706.02808.pdf)), there might be a way to use [how JAX generates random keys](https://jax.readthedocs.io/en/latest/jax.random.html#advanced) to generate a Halton sequence but its too advanced for me now. I'll use an extended HMC state for now and maybe in the future this can be improved.

**Comment by rlouf** [comment]
That works too! I know the PRNG algorithm JAX uses pretty well and there's no easy way to engineer an int sequence from it; plus since the keys are provided by the user you would get no guarantee on the ordering. 

Do we need this sequence only for adaptation or also for sampling? For the former you could also do like we do for window adaptation, which is `scan`ing over a pre-generated sequence (schedule for the window adaptation). That's much cleaner than carrying an integer.

**Comment by rlouf** [comment]
Did you manage in the end?

**Comment by albcab** [comment]
Have it almost done, just some minor bugs. I've been pretty busy these past few months, but I have some time today so I think I can get it done :+1:
